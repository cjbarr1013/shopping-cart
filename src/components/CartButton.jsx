import PropTypes from 'prop-types';
import styles from '../styles/CartButton.module.css';
import cartIcon from '../assets/icons/cart-outline.svg';
import { toUSD } from '../utils/utils';

function CartButton({ subtotal = 105.67, handleClick }) {
  return (
    <button className={styles.btn} onClick={handleClick}>
      <img src={cartIcon} alt="cart outline"></img>
      <span>{toUSD.format(subtotal)}</span>
    </button>
  );
}

CartButton.propTypes = {
  subtotal: PropTypes.number.isRequired,
  handleClick: PropTypes.func.isRequired,
};

export default CartButton;
