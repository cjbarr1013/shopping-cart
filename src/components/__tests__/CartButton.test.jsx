import { vi, describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import CartButton from '../CartButton';

// Mock CSS module
vi.mock('../styles/CartButton.module.css', () => ({
  btn: 'btn',
  animate: 'animate',
  animateImg: 'animateImg',
}));

// Mock cart icon
vi.mock('../assets/icons/cart-outline.svg', () => 'mock-cart-icon');

// Mock toUSD utility
vi.mock('../utils/utils', () => ({
  toUSD: { format: (value) => `$${value.toFixed(2)}` },
}));

describe('CartButton Component', () => {
  it('renders the button with the correct subtotal and cart icon', () => {
    render(<CartButton subtotal={100} handleClick={() => {}} />);

    const button = screen.getByRole('button');
    expect(button).toBeInTheDocument();

    const img = screen.getByAltText('cart outline');
    expect(img).toBeInTheDocument();

    const subtotal = screen.getByText('$100.00');
    expect(subtotal).toBeInTheDocument();
  });

  it('calls handleClick when the button is clicked', async () => {
    const handleClick = vi.fn();
    const user = userEvent.setup();

    render(<CartButton subtotal={100} handleClick={handleClick} />);

    const button = screen.getByRole('button');
    await user.click(button);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('adds animation class when subtotal changes', () => {
    const { rerender } = render(
      <CartButton subtotal={100} handleClick={() => {}} />
    );

    const button = screen.getByRole('button');
    expect(button).not.toHaveClass(/animate/i); // Initially, no animation

    // Change subtotal to trigger animation
    rerender(<CartButton subtotal={200} handleClick={() => {}} />);
    expect(button).toHaveClass(/animate/i); // Animation class is added

    // Fast-forward the timer
    setTimeout(() => {
      expect(button).not.toHaveClass(/animate/i); // Animation class should be removed after timeout
    }, 1000);
  });

  it('does not add animation class when subtotal remains the same', () => {
    const { rerender } = render(
      <CartButton subtotal={100} handleClick={() => {}} />
    );

    const button = screen.getByRole('button');
    expect(button).not.toHaveClass(/animate/i);

    // Rerender with the same subtotal
    rerender(<CartButton subtotal={100} handleClick={() => {}} />);
    expect(button).not.toHaveClass(/animate/i);
  });

  it('matches snapshot', () => {
    const { asFragment } = render(
      <CartButton subtotal={100} handleClick={() => {}} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
