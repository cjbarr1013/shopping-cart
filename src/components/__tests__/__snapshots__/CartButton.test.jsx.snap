// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="_primary_4048ec"
  >
    Click me!
  </button>
</DocumentFragment>
`;

exports[`CartButton Component > matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="_btn_a3b042 "
  >
    <img
      alt="cart outline"
      class=""
      src="data:image/svg+xml,%3csvg%20xmlns='http://www.w3.org/2000/svg'%20viewBox='0%200%2024%2024'%3e%3cpath%20d='M17,18A2,2%200%200,1%2019,20A2,2%200%200,1%2017,22C15.89,22%2015,21.1%2015,20C15,18.89%2015.89,18%2017,18M1,2H4.27L5.21,4H20A1,1%200%200,1%2021,5C21,5.17%2020.95,5.34%2020.88,5.5L17.3,11.97C16.96,12.58%2016.3,13%2015.55,13H8.1L7.2,14.63L7.17,14.75A0.25,0.25%200%200,0%207.42,15H19V17H7C5.89,17%205,16.1%205,15C5,14.65%205.09,14.32%205.24,14.04L6.6,11.59L3,4H1V2M7,18A2,2%200%200,1%209,20A2,2%200%200,1%207,22C5.89,22%205,21.1%205,20C5,18.89%205.89,18%207,18M16,11L18.78,6H6.14L8.5,11H16Z'%20/%3e%3c/svg%3e"
    />
    <span>
      $100.00
    </span>
  </button>
</DocumentFragment>
`;
